File: backend/app/routers/items.py
"""资料项路由。"""
from pathlib import Path
from typing import List
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.config import settings
from app.database import get_db
from app.models import Item, File
from app.schemas import (
    ItemCreate, ItemUpdate, ItemResponse, ItemListResponse, FileInItem,
    ConfirmRequest, RejectRequest
)
from app.services import item_service

router = APIRouter(tags=["items"])


def _to_response(item: Item) -> ItemResponse:
    # 修 C1 (REVIEW-TRACK1 C1)：promote_available 从 item.promote_available 拿
    # （add_item 路由临时挂的属性；其他路由默认 False）
    promote_available = getattr(item, "promote_available", False)
    return ItemResponse(
        id=item.id,
        seq=item.seq,
        name=item.name,
        description=item.description,
        pages=item.pages,
        status=item.status,
        rejected_note=item.rejected_note,
        confirmed_at=item.confirmed_at,
        is_extension=item.is_extension,
        files=[
            FileInItem(
                id=f.id,
                filename=f.filename,
                filesize=f.filesize,
                is_pdf=f.is_pdf,
                is_primary=f.is_primary,
                uploaded_at=f.uploaded_at,
            )
            for f in item.files
        ],
        promote_available=promote_available,
    )


@router.get("/api/projects/{project_id}/items", response_model=ItemListResponse)
def list_items(project_id: str, db: Session = Depends(get_db)):
    items = db.query(Item).filter(Item.project_id == project_id).order_by(Item.seq).all()
    # 修 I-4：用 item_id_orphan 列直接 SQL 过滤（不再走磁盘 — 修 TOCTOU 风险）
    # item_id_orphan 在 T-06 之后由 ingest_path 写入时填充，存当前 project_id
    unclaimed_in_project = (
        db.query(File)
        .filter(File.item_id == "", File.item_id_orphan == project_id)
        .all()
    )
    return ItemListResponse(
        project_id=project_id,
        items=[_to_response(i) for i in items],
        unclaimed=[
            FileInItem(
                id=u.id, filename=u.filename, filesize=u.filesize,
                is_pdf=u.is_pdf, is_primary=u.is_primary, uploaded_at=u.uploaded_at,
            )
            for u in unclaimed_in_project
        ],
    )


@router.post("/api/projects/{project_id}/items", response_model=ItemResponse, status_code=201)
def add_item(project_id: str, payload: ItemCreate, db: Session = Depends(get_db)):
    item, promote_available = item_service.add_item(
        db, project_id, payload.name, payload.description,
    )
    # 修 C1 (REVIEW-TRACK1 C1)：promote_available 字段已写进 ItemResponse schema。
    # 这里需要把 promote_available 挂到 item 上（不是 SQL 列，只是临时属性），
    # 让 Pydantic 序列化时能取到。
    item.promote_available = promote_available  # type: ignore[attr-defined]
    return _to_response(item)


@router.patch("/api/items/{item_id}", response_model=ItemResponse)
def update_item(item_id: str, payload: ItemUpdate, db: Session = Depends(get_db)):
    item = db.query(Item).filter(Item.id == item_id).first()
    if not item:
        raise HTTPException(404, "item 不存在")
    for k, v in payload.model_dump(exclude_unset=True).items():
        setattr(item, k, v)
    db.commit()
    db.refresh(item)
    return _to_response(item)


@router.delete("/api/items/{item_id}", status_code=204)
def delete_item(item_id: str, db: Session = Depends(get_db)):
    item = db.query(Item).filter(Item.id == item_id).first()
    if not item:
        raise HTTPException(404, "item 不存在")
    db.delete(item)
    db.commit()


@router.post("/api/items/{item_id}/confirm", response_model=ItemResponse)
def confirm_item(item_id: str, payload: ConfirmRequest = ConfirmRequest(), db: Session = Depends(get_db)):
    try:
        item = item_service.confirm_item(db, item_id, payload.primary_file_id)
    except LookupError as e:
        raise HTTPException(404, str(e))
    except ValueError as e:
        raise HTTPException(409, str(e))
    return _to_response(item)


@router.post("/api/items/{item_id}/reject", response_model=ItemResponse)
def reject_item(item_id: str, payload: RejectRequest, db: Session = Depends(get_db)):
    try:
        item = item_service.reject_item(db, item_id, payload.note)
    except LookupError as e:
        raise HTTPException(404, str(e))
    except ValueError as e:
        raise HTTPException(409, str(e))
    return _to_response(item)


@router.post("/api/items/{item_id}/reset", response_model=ItemResponse)
def reset_item(item_id: str, db: Session = Depends(get_db)):
    try:
        item = item_service.reset_item(db, item_id)
    except LookupError as e:
        raise HTTPException(404, str(e))
    return _to_response(item)

File: frontend/src/api/items.ts
/**
 *资料项（Item）API封装。
 *
 *端点清单（与 backend/app/routers/items.py 对齐）：
 * - GET /api/projects/{projectId}/items
 * - POST /api/projects/{projectId}/items （新增项 + 可能触发推广提示）
 * - PATCH /api/items/{itemId} （仅用于编辑 name/description/pages，**禁止**改 status）
 * - DELETE /api/items/{itemId}
 * - POST /api/items/{itemId}/confirm （uploaded → confirmed）
 * - POST /api/items/{itemId}/reject （uploaded → rejected，需 note）
 * - POST /api/items/{itemId}/reset （rejected/confirmed-with-no-files → pending）
 *
 *状态机（SPEC §2.1）：
 * pending → uploaded → confirmed
 * └→ rejected → pending
 *
 * 前端契约：
 * -任何 status变更 MUST走 confirm / reject / reset端点，**禁止**直接 PATCH status
 * - 见本文件 `patchItem()` 的注释
 */
import { apiClient } from './client';
import type { Item, ItemListResponse, ItemStatus, TemplateItem } from '../types';

/** 新增项的请求体 */
export interface AddItemPayload {
 name: string;
 description?: string | null;
 pages?: number | null;
}

/** 新增项的响应（带 promote提示） */
export interface AddItemResponse extends Item {
 promote_available?: boolean;
 promote_prompt?: {
 available: boolean;
 message: string;
 preview: TemplateItem;
 };
}

/** 编辑项的请求体（PATCH /api/items/{id}，不包含 status） */
export interface UpdateItemPayload {
 name?: string;
 description?: string | null;
 pages?: number | null;
}

/**复核通过的请求体 */
export interface ConfirmPayload {
 /** 主文件 ID；不传则后端自动选第一个 PDF */
 primary_file_id?: string | null;
}

/**驳回的请求体 */
export interface RejectPayload {
 /**驳回备注（必填，后端 RejectRequest强制 min_length=1） */
 note: string;
}

/**
 * 列出一个项目下所有 item + 未认领文件。
 * 注意：响应包含 `unclaimed`，但实际归属逻辑在 ProjectDetail 里通过 ItemListResponse一起拿到。
 */
export async function listItems(projectId: string): Promise<ItemListResponse> {
  const { data } = await apiClient.get<ItemListResponse>(
    `/projects/${encodeURIComponent(projectId)}/items`,
  );
  return data;
}

/**
 * 新增一项（项目级扩展项）。后端返回201 + ItemResponse，可能带 promote_available标志。
 */
export async function addItem(
  projectId: string,
  payload: AddItemPayload,
): Promise<AddItemResponse> {
  const { data } = await apiClient.post<AddItemResponse>(
    `/projects/${encodeURIComponent(projectId)}/items`,
    payload,
  );
  return data;
}

/**
 * 编辑项的元信息（name/description/pages）。
 *
 * ⚠️ SPEC-ST-1 / SPEC-ST-2 / SPEC-ST-5：前端**禁止**通过此端点改 status。
 * 如需状态变更，必须走 confirm / reject / reset专用端点。
 */
export async function updateItem(
  itemId: string,
  payload: UpdateItemPayload,
): Promise<Item> {
  const { data } = await apiClient.patch<Item>(
    `/items/${encodeURIComponent(itemId)}`,
    payload,
  );
  return data;
}

/** 删除一个项（级联文件）。 */
export async function deleteItem(itemId: string): Promise<void> {
  await apiClient.delete(`/items/${encodeURIComponent(itemId)}`);
}

/**
 *复核通过（uploaded → confirmed）。
 *
 * @param itemId资料项 ID
 * @param primaryFileId 主文件 ID；不传则后端自动选第一个
 */
export async function confirmItem(
  itemId: string,
  primaryFileId?: string | null,
): Promise<Item> {
  const payload: ConfirmPayload = {};
  if (primaryFileId) payload.primary_file_id = primaryFileId;
  const { data } = await apiClient.post<Item>(
    `/items/${encodeURIComponent(itemId)}/confirm`,
    payload,
  );
  return data;
}

/**
 *驳回（uploaded → rejected）。
 * 必须提供驳回备注。
 */
export async function rejectItem(itemId: string, note: string): Promise<Item> {
  const { data } = await apiClient.post<Item>(
    `/items/${encodeURIComponent(itemId)}/reject`,
    { note },
  );
  return data;
}

/**
 * 重置为 pending（rejected / confirmed-with-no-files → pending）。
 * 通常用于驳回后允许重新上传，或管理员强制重置。
 */
export async function resetItem(itemId: string): Promise<Item> {
  const { data } = await apiClient.post<Item>(
    `/items/${encodeURIComponent(itemId)}/reset`,
  );
  return data;
}

/** 类型守卫：合法的 ItemStatus */
export function isItemStatus(s: string): s is ItemStatus {
 return s === 'pending' || s === 'uploaded' || s === 'confirmed' || s === 'rejected';
}
